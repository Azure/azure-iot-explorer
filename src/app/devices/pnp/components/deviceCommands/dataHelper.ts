/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';
import { JsonSchemaAdaptor } from '../../../../shared/utils/jsonSchemaAdaptor';
import { CommandSchema } from './deviceCommandsPerInterfacePerCommand';

interface DeviceCommandsWithSchema {
    commandSchemas: CommandSchema[];
}

export const getDeviceCommandPairs = (modelDefinition: ModelDefinition): DeviceCommandsWithSchema => {
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(modelDefinition);
    const commands = jsonSchemaAdaptor.getCommands();
    return {
        // tslint:disable-next-line: no-any
        commandSchemas: commands.map((command: any) => ({
            commandModelDefinition: command,
            parsedSchema: jsonSchemaAdaptor.parseInterfaceCommandToJsonSchema(command),
        }))
    };
};
