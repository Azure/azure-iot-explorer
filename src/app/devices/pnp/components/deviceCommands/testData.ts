/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';
import { PnpStateInterface, pnpStateInitial } from '../../state';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { REPOSITORY_LOCATION_TYPE } from '../../../../constants/repositoryLocationTypes';

const interfaceId = 'urn:contoso:com:environmentalsensor;1';
/* tslint:disable */
export const testModelDefinition: ModelDefinition = {
    "@id": interfaceId,
    "@type": "Interface",
    "displayName": "Digital Twin",
    "contents": [
        {
            "@type": "Command",
            "description": "This command will begin blinking the LED for given time interval.",
            "name": "blink",
            "request": {
                "name": "blinkRequest",
                "schema": "long"
            },
            "response": {
                "name": "blinkResponse",
                "schema": "string"
            },
            "commandType": "synchronous"
        }
    ],
    "@context": "https://azureiot.com/v1/contexts/Interface.json"
};

export const pnpStateWithTestData: PnpStateInterface = {
    ...pnpStateInitial(),
    modelDefinitionWithSource: {
        synchronizationStatus: SynchronizationStatus.fetched,
        payload: {
            isModelValid: true,
            modelDefinition: testModelDefinition,
            source: REPOSITORY_LOCATION_TYPE.Public
        }
    }
}