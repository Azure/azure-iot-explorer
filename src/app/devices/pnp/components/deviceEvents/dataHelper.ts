/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition, TelemetryContent } from '../../../../api/models/modelDefinition';
import { JsonSchemaAdaptor } from '../../../../shared/utils/jsonSchemaAdaptor';
import { ParsedJsonSchema } from '../../../../api/models/interfaceJsonParserOutput';

export interface TelemetrySchema {
    parsedSchema: ParsedJsonSchema;
    telemetryModelDefinition: TelemetryContent;
}

export const getDeviceTelemetry = (modelDefinition: ModelDefinition): TelemetrySchema[] => {
    if (!modelDefinition) {
        return [];
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(modelDefinition);
    const telemetryContents = jsonSchemaAdaptor.getTelemetry();
    return telemetryContents.map(telemetry => ({
        parsedSchema: jsonSchemaAdaptor.parseInterfaceTelemetryToJsonSchema(telemetry),
        telemetryModelDefinition: telemetry
    }));
};
