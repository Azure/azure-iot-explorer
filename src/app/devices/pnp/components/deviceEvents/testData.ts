/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';

const interfaceId = 'urn:azureiot:samplemodel;1';
export const testComponentName = 'environmentalSensor';
/* tslint:disable */
export const testModelDefinition: ModelDefinition = {
    "@id": interfaceId,
    "@type": "Interface",
    "displayName": "Digital Twin",
    "contents": [
        {
            "@type": "Telemetry",
            "name": "humid",
            "displayName": "Temperature",
            "description": "Current temperature on the device",
            "schema": "double"
        },
    ],
    "@context": "https://azureiot.com/v1/contexts/Interface.json"
}
/* tslint:enable */
