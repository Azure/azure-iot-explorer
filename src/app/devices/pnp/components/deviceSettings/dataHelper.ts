/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { JsonSchemaAdaptor } from '../../../../shared/utils/jsonSchemaAdaptor';
import { ModelDefinition, PropertyContent } from '../../../../api/models/modelDefinition';
import { ParsedJsonSchema } from '../../../../api/models/interfaceJsonParserOutput';
import { ModuleTwin } from '../../../../api/models/moduleTwin';
import { Twin } from '../../../../api/models/device';
import { DEFAULT_COMPONENT_FOR_DIGITAL_TWIN } from '../../../../constants/devices';

export interface ReportedSection {
    value?: boolean | string | number | object;
    ac?: number;
    ad?: string;
    av?: number;
}

export interface TwinWithSchema {
    settingModelDefinition: PropertyContent;
    settingSchema: ParsedJsonSchema;
    desiredValue?: boolean | string | number | object;
    reportedSection?: ReportedSection;
}

export const generateTwinSchemaAndInterfaceTuple = (model: ModelDefinition, twin: Twin | ModuleTwin, componentName: string): TwinWithSchema[] => {
    if (!model) {
        return [];
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(model);
    const writablePropertyModelDefinitions = jsonSchemaAdaptor.getWritableProperties();

    const settings = writablePropertyModelDefinitions
        .map(writablePropertyModelDefinition => {
            return {
                desiredValue: getDesiredPropertiesForSpecficComponent(twin, componentName)?.[writablePropertyModelDefinition.name],
                reportedSection: getReportedPropertiesForSpecficComponent(twin, componentName)?.[writablePropertyModelDefinition.name],
                settingModelDefinition: writablePropertyModelDefinition,
                settingSchema: jsonSchemaAdaptor.parseInterfacePropertyToJsonSchema(writablePropertyModelDefinition)
            };
        });

    return settings;
};

export const getReportedPropertiesForSpecficComponent = (twin: Twin | ModuleTwin, componentName: string) => {
    const reportedTwin = twin?.properties?.reported;
    return componentName === DEFAULT_COMPONENT_FOR_DIGITAL_TWIN ?
        reportedTwin : (reportedTwin as any)?.[componentName]; // tslint:disable-line: no-any
};

export const getDesiredPropertiesForSpecficComponent = (twin: Twin | ModuleTwin, componentName: string) => {
    const desiredTwin = twin?.properties?.desired;
    return componentName === DEFAULT_COMPONENT_FOR_DIGITAL_TWIN ?
        desiredTwin : (desiredTwin as any)?.[componentName]; // tslint:disable-line: no-any
};
