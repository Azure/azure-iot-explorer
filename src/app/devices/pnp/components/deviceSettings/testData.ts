/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { ModelDefinition } from '../../../../api/models/modelDefinition';
import { PnpStateInterface, pnpStateInitial } from '../../state';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { REPOSITORY_LOCATION_TYPE } from '../../../../constants/repositoryLocationTypes';

const interfaceId = 'urn:azureiot:samplemodel;1';
export const testComponentName = 'environmentalSensor';
/* tslint:disable */
export const testTwin: any = {
    "deviceId": "testDevice",
    properties: {
        desired: {
            "brightness": 5678, // root level brightness,
            environmentalSensor: {
                "__t": "c",
                "brightness": 456,  // component level brightness
            }
        },
        reported: {
            "brightness": {
                "value": 1234,
                "ac": 200,
                "ad": "success",
                "av": 2
            },
            environmentalSensor: {
                "__t": "c",
                "brightness": {
                    "value": 123,
                    "ac": 200,
                    "ad": "success",
                    "av": 2
                }
            }
        }
    }
};

const brightnessProperty = {
    '@type': 'Property',
    displayName: 'Brightness Level',
    description: 'The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)',
    name: 'brightness',
    writable: true,
    schema: 'long'
};

const sampleSenmanticProperty = {
    '@type': [
        "Property",
        "SemanticType/Brightness Level"
        ],
    displayName: 'Brightness Level',
    description: 'The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)',
    name: 'brightness',
    writable: true,
    schema: 'long'
};

export const testModelDefinition: ModelDefinition = {
    '@id': interfaceId,
    '@type': 'Interface',
    '@context': 'https://azureiot.com/v1/contexts/Interface.json',
    displayName: 'Environmental Sensor',
    description: 'Provides functionality to report temperature, humidity. Provides telemetry, commands and read-write properties',
    comment: 'Requires temperature and humidity sensors.',
    contents: [ brightnessProperty, sampleSenmanticProperty ]
}

export const pnpStateWithTestData: PnpStateInterface = {
    ...pnpStateInitial(),
    modelDefinitionWithSource: {
        synchronizationStatus: SynchronizationStatus.fetched,
        payload: {
            modelDefinition: testModelDefinition,
            source: REPOSITORY_LOCATION_TYPE.Local,
            isModelValid: true
        },
    },
    twin: {
        payload: testTwin,
        synchronizationStatus: SynchronizationStatus.fetched
    }
}
