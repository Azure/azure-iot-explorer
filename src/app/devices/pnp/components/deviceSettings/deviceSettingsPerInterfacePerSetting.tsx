/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { Label, Stack, ActionButton, IconButton } from '@fluentui/react';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { InterfaceDetailCard } from '../../../../constants/iconNames';
import { ComplexReportedFormPanel } from '../../../shared/components/complexReportedFormPanel';
import { RenderSimplyTypeValue } from '../../../shared/components/simpleReportedSection';
import { ErrorBoundary } from '../../../shared/components/errorBoundary';
import { getLocalizedData } from '../../../../api/dataTransforms/modelDefinitionTransform';
import { SemanticUnit } from '../../../../shared/units/components/semanticUnit';
import { DataForm } from '../../../shared/components/dataForm';
import { TwinWithSchema } from './dataHelper';
import { DEFAULT_COMPONENT_FOR_DIGITAL_TWIN } from '../../../../constants/devices';
import { getSchemaType, isSchemaSimpleType } from '../../../../shared/utils/jsonSchemaAdaptor';
import { Twin } from '../../../../api/models/device';
import { ModuleTwin } from '../../../../api/models/moduleTwin';
import { EnumSchema } from '../../../../api/models/modelDefinition';

export interface DeviceSettingDataProps extends TwinWithSchema {
    collapsed: boolean;
    deviceId: string;
    moduleId: string;
    interfaceId: string;
    componentName: string;
}

export interface DeviceSettingDispatchProps {
    handleCollapseToggle: () => void;
    handleOverlayToggle: () => void;
    patchTwin: (parameters: Partial<Twin>) => void;
}

export const DeviceSettingsPerInterfacePerSetting: React.FC<DeviceSettingDataProps & DeviceSettingDispatchProps> = (props: DeviceSettingDataProps & DeviceSettingDispatchProps) => {
    const { t } = useTranslation();

    const { settingSchema, settingModelDefinition, collapsed, reportedSection, handleOverlayToggle } = props;
    const [ showReportedValuePanel, setShowReportedValuePanel ] = React.useState<boolean>(false);

    const createCollapsedSummary = () => {
        return (
            <header className={`flex-grid-row item-summary ${collapsed ? '' : 'item-summary-uncollapsed'}`} onClick={props.handleCollapseToggle}>
                {renderPropertyName()}
                {renderPropertySchema()}
                {renderPropertyUnit()}
                {renderReportedValueAndMetadata()}
                {renderCollapseButton()}
            </header>
        );
    };

    const renderPropertyName = () => {
        const ariaLabel = t(ResourceKeys.deviceSettings.columns.name);
        let displayName = getLocalizedData(settingModelDefinition.displayName);
        displayName = displayName ? displayName : '--';
        let description = getLocalizedData(settingModelDefinition.description);
        description = description ? description : '--';
        return <div className="col-sm3"><Label aria-label={ariaLabel}>{settingModelDefinition.name} ({displayName} / {description})</Label></div>;
    };

    const renderPropertySchema = () => {
        const ariaLabel = t(ResourceKeys.deviceProperties.columns.schema);
        const schemaType = getSchemaType(settingModelDefinition.schema);
        return <div className="col-sm2"><Label aria-label={ariaLabel}>{schemaType}</Label></div>;
    };

    const renderPropertyUnit = () => {
        return (
            <div className="col-sm2">
                <Label aria-label={t(ResourceKeys.deviceProperties.columns.unit)}>
                    <SemanticUnit unitHost={settingModelDefinition} />
                </Label>
            </div>);
    };

    // tslint:disable-next-line: cyclomatic-complexity
    const renderReportedValueAndMetadata = () => {
        const ariaLabel = t(ResourceKeys.deviceSettings.columns.reportedValue);
        return (
            <div className="column-value-text col-sm4" aria-label={ariaLabel}>
                <Stack>
                    <Stack.Item align="start" style={{maxWidth: '40%', minWidth: 20}}>
                        {renderReportedValue()}
                    </Stack.Item>
                    <Stack.Item align="start">
                        {t(ResourceKeys.deviceSettings.ackStatus.code, {code: reportedSection?.ac || '--'})}
                    </Stack.Item>
                    <Stack.Item align="start">
                        {t(ResourceKeys.deviceSettings.ackStatus.description, {description: reportedSection?.ad || '--'})}
                    </Stack.Item>
                    <Stack.Item align="start">
                        {t(ResourceKeys.deviceSettings.ackStatus.version, {version: reportedSection?.av || '--'})}
                    </Stack.Item>
                </Stack>
            </div>
        );
    };

    // tslint:disable-next-line:cyclomatic-complexity
    const renderReportedValue = () => {
        const reportedValue = reportedSection?.value !== undefined ? reportedSection.value : reportedSection;
        let displayValue = (settingModelDefinition?.schema as EnumSchema)?.enumValues?.find(item => item.enumValue === reportedValue)?.displayName || reportedValue;
        // tslint:disable-next-line:no-any
        displayValue = (displayValue as any)?.en || displayValue;
        return (
            <ErrorBoundary error={t(ResourceKeys.errorBoundary.text)}>
                {
                    settingSchema && isSchemaSimpleType(settingModelDefinition.schema, settingSchema.$ref) ?
                        RenderSimplyTypeValue(
                            reportedValue,
                            settingSchema,
                            displayValue,
                            t(ResourceKeys.deviceSettings.columns.error)) :
                        reportedValue ?
                            <ActionButton
                                className="column-value-button"
                                ariaDescription={t(ResourceKeys.deviceSettings.command.openReportedValuePanel)}
                                onClick={onViewReportedValue}
                            >
                                {t(ResourceKeys.deviceSettings.command.openReportedValuePanel)}
                            </ActionButton> : <Label>--</Label>
                }
            </ErrorBoundary>
        );
    };

    const renderCollapseButton = () => {
        return (
        <div className="col-sm1">
            <IconButton
                title={t(collapsed ? ResourceKeys.deviceSettings.command.expand : ResourceKeys.deviceSettings.command.collapse)}
                iconProps={{iconName: collapsed ? InterfaceDetailCard.OPEN : InterfaceDetailCard.CLOSE}}
            />
        </div>);
    };

    const createUncollapsedCard = () => {
        return (
            <section className={collapsed ? 'item-detail' : 'item-detail item-detail-uncollapsed'}>
                {!collapsed && createForm()}
            </section>
        );
    };

    const createReportedValuePanel = () => {
        return (
            <div role="dialog">
                {showReportedValuePanel &&
                    <ComplexReportedFormPanel
                        schema={settingSchema}
                        modelDefinition={settingModelDefinition}
                        showPanel={showReportedValuePanel}
                        formData={reportedSection?.value}
                        handleDismiss={handleDismissViewReportedPanel}
                    />}
            </div>
        );
    };

    const onViewReportedValue = (event: React.MouseEvent<HTMLElement>) => {
        setShowReportedValuePanel(true);
        handleOverlayToggle();
        if (event) {
            event.stopPropagation();
        }
    };

    const handleDismissViewReportedPanel = () => {
        setShowReportedValuePanel(false);
        handleOverlayToggle();
    };

    const createForm = () => {
        const { desiredValue } = props;
        return (
            <DataForm
                buttonText={ResourceKeys.deviceSettings.command.submit}
                formData={desiredValue !== undefined ? desiredValue : reportedSection?.value}
                settingSchema={settingSchema}
                handleSave={onSubmit}
                schema={getSchemaType(settingModelDefinition.schema)}
            />
        );
    };

    const createSettingsPayload = (twin: boolean | number | string | object): Partial<Twin | ModuleTwin> => {
        const { componentName, deviceId, moduleId } = props;
        const desired = {} as any; // tslint:disable-line: no-any
        if (componentName === DEFAULT_COMPONENT_FOR_DIGITAL_TWIN) {
            desired[settingModelDefinition.name] = twin;
        }
        else {
            const componentValue = {__t: 'c'} as any; // tslint:disable-line: no-any
            componentValue[settingModelDefinition.name] = twin;
            desired[componentName] = componentValue;
        }

        return {
            deviceId,
            moduleId,
            properties: {
                desired
            }
        };
    };

    const onSubmit = (twin: boolean | number | string | object) => () => {
        props.patchTwin(createSettingsPayload(twin));
    };

    return (
        <article className="list-item" role="listitem">
            <ErrorBoundary error={t(ResourceKeys.errorBoundary.text)}>
                {createCollapsedSummary()}
                {createUncollapsedCard()}
                {createReportedValuePanel()}
            </ErrorBoundary>
        </article>
    );
};
