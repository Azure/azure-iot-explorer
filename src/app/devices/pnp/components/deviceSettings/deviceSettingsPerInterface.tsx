/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { IconButton, Overlay } from '@fluentui/react';
import { DeviceSettingsPerInterfacePerSetting } from './deviceSettingsPerInterfacePerSetting';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { InterfaceDetailCard } from '../../../../constants/iconNames';
import { LabelWithTooltip } from '../../../../shared/components/labelWithTooltip';
import { TwinWithSchema } from './dataHelper';
import '../../../../css/_devicePnpDetailList.scss';
import { Twin } from '../../../../api/models/device';

export interface DeviceSettingDataProps {
    deviceId: string;
    moduleId: string;
    interfaceId: string;
    componentName: string;
    twinWithSchema: TwinWithSchema[];
}

export interface DeviceSettingDispatchProps {
    patchTwin: (parameters: Twin) => void;
}

export interface DeviceSettingState {
    collapseMap: Map<number, boolean>;
    allCollapsed: boolean;
    showOverlay: boolean;
}

// tslint:disable-next-line: cyclomatic-complexity
export const DeviceSettingsPerInterface: React.FC<DeviceSettingDataProps & DeviceSettingDispatchProps> = (props: DeviceSettingDataProps & DeviceSettingDispatchProps) => {
    const { t } = useTranslation();
    const { twinWithSchema } = props;
    const initialCollapseMap = new Map();
    for (let index = 0; index < twinWithSchema.length; index ++) {
        initialCollapseMap.set(index, false);
    }
    const [ allCollapsed, setAllCollapsed ] = React.useState<boolean>(false);
    const [ collapseMap, setCollapseMap ] = React.useState(initialCollapseMap);
    const [ showOverlay, setShowOverlay ] = React.useState<boolean>(false);

    const renderCollapseAllButton = () => {
        return (
            <div className="col-sm1 collapse-button">
                <IconButton
                    iconProps={{iconName: allCollapsed ? InterfaceDetailCard.OPEN : InterfaceDetailCard.CLOSE}}
                    ariaLabel={allCollapsed ?
                        t(ResourceKeys.deviceSettings.command.expandAll) :
                        t(ResourceKeys.deviceSettings.command.collapseAll)}
                    onClick={onToggleCollapseAll}
                    title={t(allCollapsed ? ResourceKeys.deviceSettings.command.expandAll : ResourceKeys.deviceSettings.command.collapseAll)}
                />
            </div>
        );
    };

    const onToggleCollapseAll = () => {
        const newCollapseMap = new Map();
        for (let index = 0; index < collapseMap.size; index ++) {
            newCollapseMap.set(index, !allCollapsed);
        }
        setCollapseMap(newCollapseMap);
        setAllCollapsed(!allCollapsed);
    };

    const handleCollapseToggle = (index: number) => () => {
        const newCollapseMap = new Map(collapseMap);
        newCollapseMap.set(index, !newCollapseMap.get(index));
        setCollapseMap(newCollapseMap);
    };

    const handleOverlayToggle = () => {
        setShowOverlay(!showOverlay);
    };

    const settings = twinWithSchema && twinWithSchema.map((tuple, index) => (
        <DeviceSettingsPerInterfacePerSetting
            key={index}
            {...tuple}
            {...props}
            collapsed={collapseMap.get(index)}
            handleCollapseToggle={handleCollapseToggle(index)}
            handleOverlayToggle={handleOverlayToggle}
        />
    ));

    return (
        <div className="pnp-detail-list pnp-properties scrollable-lg">
            <div className="list-header flex-grid-row">
                <span className="col-sm3">{t(ResourceKeys.deviceSettings.columns.name)}</span>
                <span className="col-sm2">{t(ResourceKeys.deviceSettings.columns.schema)}</span>
                <span className="col-sm2">{t(ResourceKeys.deviceSettings.columns.unit)}</span>
                <span className="col-sm4 reported-value">
                    <LabelWithTooltip tooltipText={t(ResourceKeys.deviceSettings.columns.reportedValueTooltip)}>
                        {t(ResourceKeys.deviceSettings.columns.reportedValue)}
                    </LabelWithTooltip>
                </span>
                {renderCollapseAllButton()}
            </div>
            <section role={twinWithSchema && twinWithSchema.length === 0 ? 'main' : 'list'} className="list-content">
                {settings}
            </section>
            {showOverlay && <Overlay/>}
        </div>
    );
};
