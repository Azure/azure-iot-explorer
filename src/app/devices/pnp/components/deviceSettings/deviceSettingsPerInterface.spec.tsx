/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import { IconButton, Overlay } from '@fluentui/react';
import { DeviceSettingsPerInterface, DeviceSettingDataProps, DeviceSettingDispatchProps } from './deviceSettingsPerInterface';
import { DeviceSettingsPerInterfacePerSetting } from './deviceSettingsPerInterfacePerSetting';
import { generateTwinSchemaAndInterfaceTuple } from './dataHelper';
import { testModelDefinition, testTwin, testComponentName } from './testData';

describe('components/devices/deviceSettingsPerInterface', () => {
    const twinWithSchema = generateTwinSchemaAndInterfaceTuple(testModelDefinition, testTwin, testComponentName);
    const deviceSettingsProps: DeviceSettingDataProps = {
        componentName: testComponentName,
        deviceId: 'testDevice',
        interfaceId: 'urn:contoso:com:EnvironmentalSensor;1',
        twinWithSchema,
        moduleId: ''
    };

    const deviceSettingsDispatchProps: DeviceSettingDispatchProps = {
        patchTwin: jest.fn()
    };

    const getComponent = (overrides = {}) => {
        const props = {
            ...deviceSettingsProps,
            ...deviceSettingsDispatchProps,
            ...overrides
        };

        return <DeviceSettingsPerInterface {...props} />;
    };

    it('matches snapshot', () => {
        expect(shallow(getComponent())).toMatchSnapshot();
    });

    it('toggles collapsed', () => {
        const wrapper = mount(getComponent());
        const iconButton = wrapper.find('.collapse-button').find(IconButton).first();
        expect(iconButton.props().title).toEqual('deviceSettings.command.collapseAll');

        act(() => iconButton.props().onClick(undefined));
        wrapper.update();
        const updatedIconButton = wrapper.find('.collapse-button').find(IconButton).first();
        expect(updatedIconButton.props().title).toEqual('deviceSettings.command.expandAll');
    });

    it('executes handle toggle from child', () => {
        const wrapper = mount(getComponent());

        expect(wrapper.find(DeviceSettingsPerInterfacePerSetting).first().props().collapsed).toBeFalsy();

        act(() => wrapper.find(DeviceSettingsPerInterfacePerSetting).first().props().handleCollapseToggle());
        wrapper.update();
        expect(wrapper.find(DeviceSettingsPerInterfacePerSetting).first().props().collapsed).toBeTruthy();
    });

    it('shows overlay', () => {
        const wrapper = mount(getComponent());
        expect(wrapper.find(Overlay)).toEqual({});

        act(() => wrapper.find(DeviceSettingsPerInterfacePerSetting).first().props().handleOverlayToggle());
        wrapper.update();
        expect(wrapper.find(Overlay)).toBeDefined();
    });
});
