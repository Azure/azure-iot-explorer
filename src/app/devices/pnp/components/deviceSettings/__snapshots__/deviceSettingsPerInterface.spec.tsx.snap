// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/deviceSettingsPerInterface matches snapshot 1`] = `
<div
  className="pnp-detail-list pnp-properties scrollable-lg"
>
  <div
    className="list-header flex-grid-row"
  >
    <span
      className="col-sm3"
    >
      deviceSettings.columns.name
    </span>
    <span
      className="col-sm2"
    >
      deviceSettings.columns.schema
    </span>
    <span
      className="col-sm2"
    >
      deviceSettings.columns.unit
    </span>
    <span
      className="col-sm4 reported-value"
    >
      <LabelWithTooltip
        tooltipText="deviceSettings.columns.reportedValueTooltip"
      >
        deviceSettings.columns.reportedValue
      </LabelWithTooltip>
    </span>
    <div
      className="col-sm1 collapse-button"
    >
      <CustomizedIconButton
        ariaLabel="deviceSettings.command.collapseAll"
        iconProps={
          Object {
            "iconName": "ChevronUp",
          }
        }
        onClick={[Function]}
        title="deviceSettings.command.collapseAll"
      />
    </div>
  </div>
  <section
    className="list-content"
    role="list"
  >
    <DeviceSettingsPerInterfacePerSetting
      collapsed={false}
      componentName="environmentalSensor"
      desiredValue={456}
      deviceId="testDevice"
      handleCollapseToggle={[Function]}
      handleOverlayToggle={[Function]}
      interfaceId="urn:contoso:com:EnvironmentalSensor;1"
      key="0"
      moduleId=""
      patchTwin={[MockFunction]}
      reportedSection={
        Object {
          "ac": 200,
          "ad": "success",
          "av": 2,
          "value": 123,
        }
      }
      settingModelDefinition={
        Object {
          "@type": "Property",
          "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
          "displayName": "Brightness Level",
          "name": "brightness",
          "schema": "long",
          "writable": true,
        }
      }
      settingSchema={
        Object {
          "definitions": Object {},
          "required": Array [],
          "title": "brightness",
          "type": Array [
            "number",
            "null",
          ],
        }
      }
      twinWithSchema={
        Array [
          Object {
            "desiredValue": 456,
            "reportedSection": Object {
              "ac": 200,
              "ad": "success",
              "av": 2,
              "value": 123,
            },
            "settingModelDefinition": Object {
              "@type": "Property",
              "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
              "displayName": "Brightness Level",
              "name": "brightness",
              "schema": "long",
              "writable": true,
            },
            "settingSchema": Object {
              "definitions": Object {},
              "required": Array [],
              "title": "brightness",
              "type": Array [
                "number",
                "null",
              ],
            },
          },
          Object {
            "desiredValue": 456,
            "reportedSection": Object {
              "ac": 200,
              "ad": "success",
              "av": 2,
              "value": 123,
            },
            "settingModelDefinition": Object {
              "@type": Array [
                "Property",
                "SemanticType/Brightness Level",
              ],
              "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
              "displayName": "Brightness Level",
              "name": "brightness",
              "schema": "long",
              "writable": true,
            },
            "settingSchema": Object {
              "definitions": Object {},
              "required": Array [],
              "title": "brightness",
              "type": Array [
                "number",
                "null",
              ],
            },
          },
        ]
      }
    />
    <DeviceSettingsPerInterfacePerSetting
      collapsed={false}
      componentName="environmentalSensor"
      desiredValue={456}
      deviceId="testDevice"
      handleCollapseToggle={[Function]}
      handleOverlayToggle={[Function]}
      interfaceId="urn:contoso:com:EnvironmentalSensor;1"
      key="1"
      moduleId=""
      patchTwin={[MockFunction]}
      reportedSection={
        Object {
          "ac": 200,
          "ad": "success",
          "av": 2,
          "value": 123,
        }
      }
      settingModelDefinition={
        Object {
          "@type": Array [
            "Property",
            "SemanticType/Brightness Level",
          ],
          "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
          "displayName": "Brightness Level",
          "name": "brightness",
          "schema": "long",
          "writable": true,
        }
      }
      settingSchema={
        Object {
          "definitions": Object {},
          "required": Array [],
          "title": "brightness",
          "type": Array [
            "number",
            "null",
          ],
        }
      }
      twinWithSchema={
        Array [
          Object {
            "desiredValue": 456,
            "reportedSection": Object {
              "ac": 200,
              "ad": "success",
              "av": 2,
              "value": 123,
            },
            "settingModelDefinition": Object {
              "@type": "Property",
              "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
              "displayName": "Brightness Level",
              "name": "brightness",
              "schema": "long",
              "writable": true,
            },
            "settingSchema": Object {
              "definitions": Object {},
              "required": Array [],
              "title": "brightness",
              "type": Array [
                "number",
                "null",
              ],
            },
          },
          Object {
            "desiredValue": 456,
            "reportedSection": Object {
              "ac": 200,
              "ad": "success",
              "av": 2,
              "value": 123,
            },
            "settingModelDefinition": Object {
              "@type": Array [
                "Property",
                "SemanticType/Brightness Level",
              ],
              "description": "The brightness level for the light on the device. Can be specified as 1 (high), 2 (medium), 3 (low)",
              "displayName": "Brightness Level",
              "name": "brightness",
              "schema": "long",
              "writable": true,
            },
            "settingSchema": Object {
              "definitions": Object {},
              "required": Array [],
              "title": "brightness",
              "type": Array [
                "number",
                "null",
              ],
            },
          },
        ]
      }
    />
  </section>
</div>
`;
